lib/routes/utils/signin.js

```diff
@@ -0,0 +1,485 @@
+  const unblockCodeLen = config.signinUnblock && config.signinUnblock.codeLength || 8
+
+  return {
+
+    validators: {
+      UNBLOCK_CODE: isA.string().regex(BASE_36).length(unblockCodeLen).optional()
+    },
+
+    /**
+     * Check if the password a user entered matches the one on
+     * file for the account. If it does not, flag the account with
+     * customs. Higher level code will take care of
+     * returning an error to the user.
+     */
+    checkPassword(accountRecord, password, clientAddress) {
+      if (butil.buffersAreEqual(accountRecord.authSalt, butil.ONES)) {
+        return customs.flag(clientAddress, {
+          email: accountRecord.email,
+          errno: error.ERRNO.ACCOUNT_RESET
+        }).then(() => {
+          throw error.mustResetAccount(accountRecord.email)
+        })
+      }
+      return password.verifyHash()
+        .then(verifyHash => {
+          return db.checkPassword(accountRecord.uid, verifyHash)
+        })
+        .then(match => {
+          if (match) {
+            return match
+          }
+          return customs.flag(clientAddress, {
+            email: accountRecord.email,
+            errno: error.ERRNO.INCORRECT_PASSWORD
+          }).then(() => {
+            match
+          })
+        })
+    },
+
+    /**
+     * Check if the user is logging in with the correct email address
+     * for their account.  If not, throw an error that will help them
+     * find the right one.
+     */
+    checkEmailAddress(accountRecord, email, originalLoginEmail) {
+      // The `originalLoginEmail` param is specified in the login request if the user has
+      // changed their primary email address, and hence needed to use a different email for
+      // calculating authPW than they use for identifying their account.
+      if (originalLoginEmail) {
+        // They must have been originally trying to log in with the correct
+        // primary email address, not a secondary one.
+        if (originalLoginEmail.toLowerCase() !== accountRecord.primaryEmail.normalizedEmail) {
+          throw error.cannotLoginWithSecondaryEmail()
+        }
+
+        // If a user changes their primary email and then deletes the old primary
+        // email from their account, that email address is basically "bricked" -
+        // the user can't log in with it, and nobody can create a new account with it.
+        // The below causes us to return a special error when trying to sign in
+        // with such a "bricked" email, because otherwise the default error message
+        // would offer the user to create an account with that email, which they can't.
+        let usingValidEmail = false
+        accountRecord.emails.some((emailData) => {
+          if (emailData.normalizedEmail === originalLoginEmail.toLowerCase()) {
+            usingValidEmail = true
+            return true
+          }
+        })
+        if (! usingValidEmail) {
+          throw error.cannotLoginWithEmail()
+        }
```

Comment:
@philbooth so I started work on refactoring the tests to cover this file explicitly, and I found that (a) we don't have any existing test coverage for this "usingValidEmail" logic, and (b) I don't think it can actually ever be triggered, because of the `cannotLoginWithSecondaryEmail` condition immediately above it.  So I'm going to remove it, and score one for the Just Do The Test Cleanup While We're Here approach.

@vbudhram when you get back, can you please walk me through what this little snippet of code is for?

Reply:
@vbudhram thanks; is this already covered by the ` if (originalLoginEmail.toLowerCase() !== accountRecord.primaryEmail.normalizedEmail)` check above, or is there another edge-case here that I'm missing?

Reply:
> is this already covered

Actually, you are right, it does get covered by `if (originalLoginEmail.toLowerCase() !== accountRecord.primaryEmail.normalizedEmail)`. I think the original comment was valid at some point, but might have been missed when doing some refactoring of the code. Feel free to remove!